src/arch/x86/x86_64/src/bzimage.rs

```diff
@@ -0,0 +1,293 @@
+// https://github.com/machyve/xhyve/blob/master/src/firmware/kexec.c
+pub struct BzImage {
+    pub low_mem_size: u64,
+    pub high_mem_start: u64,
+    pub high_mem_size: u64,
+    pub rom_base: usize,
+    pub rom_size: usize,
+    pub load: usize,
+    pub entry: usize,
+}
+
+impl Default for BootParams {
+    fn default() -> Self {
+        BootParams {
+            screen_info: [0u8; 0x040],             // 0x000
+            apm_bios_info: [0u8; 0x054 - 0x040],   // 0x040
+            _pad2: [0u8; 4],                       // 0x054
+            tboot_addr: 0u64,                      // 0x058
+            ist_info: [0u8; 0x070 - 0x060],        // 0x060
+            acpi_rsdp_addr: [0u8; 0x078 - 0x070],  // 0x070
+            _pad3: [0u8; 8],                       // 0x078
+            hd0_info: [0u8; 0x090 - 0x080],        // 0x080 /* obsolete! */
+            hd1_info: [0u8; 0x0a0 - 0x090],        // 0x090 /* obsolete! */
+            sys_desc_table: [0u8; 0x0b0 - 0x0a0],  // 0x0a0 /* obsolete! */
+            olpc_ofw_header: [0u8; 0x0c0 - 0x0b0], // 0x0b0
+            ext_ramdisk_image: 0u32,               // 0x0c0
+            ext_ramdisk_size: 0u32,                // 0x0c4
+            ext_cmd_line_ptr: 0u32,                // 0x0c8
+            _pad4: [0u8; 116],                     // 0x0cc
+            edid_info: [0u8; 0x1c0 - 0x140],       // 0x140
+            efi_info: [0u8; 0x1e0 - 0x1c0],        // 0x1c0
+            alt_mem_k: 0u32,                       // 0x1e0
+            scratch: 0u32,                         // 0x1e4 /* obsolete! */
+            e820_entries: 0u8,                     // 0x1e8
+            eddbuf_entries: 0u8,                   // 0x1e9
+            edd_mbr_sig_buf_entries: 0u8,          // 0x1ea
+            kbd_status: 0u8,                       // 0x1eb
+            secure_boot: 0u8,                      // 0x1ec
+            _pad5: [0u8; 2],                       // 0x1ed
+            sentinel: 0u8,                         // 0x1ef
+            _pad6: [0u8; 1],                       // 0x1f0
+            hdr: SetupHeader::default(),           // 0x1f1
+            _pad7: [0u8; 0x290 - 0x1f1 - size_of::<SetupHeader>()],
+            edd_mbr_sig_buffer: [0u32; 16],               // 0x290
+            e820_table: [E820Entry::default(); E820_MAX], // 0x2d0
+            _pad8: [0u8; 48],                             // 0xcd0
+            eddbuf: [0u8; 0xeec - 0xd00],                 // 0xd00
+            _pad9: [0u8; 276],                            // 0xeec
+        }
+    }
+}
+
+impl BzImage {
+    pub fn load(&mut self, w: &mut print::WriteTo) -> Result<usize> {
+        // The raw pointer shit is too painful.
+        let rom = SectionReader::new(&Memory {}, self.rom_base, self.rom_size);
+        let mut header: SetupHeader = {
+            let mut buff = [0u8; size_of::<SetupHeader>()];
+            rom.pread(&mut buff, HEADER_OFFSET).unwrap();
+            unsafe { transmute(buff) }
```

Comment:
Is there a way to do this safely?

Reply:
There is a safe_transmute crate but I'm not sure about the guarantees there.

Reply:
The only current safe ways I know about are field-by-field copy, similar to our field-by-field zeroing that we did implementing Default, or writing a parser, e.g., using the `nom` crate. Both of these probably can be made to compile down to zero overhead. They are safer if we're wrong about alignment/padding guarantees.